### PR TITLE
[WIP][Plasma] Use jemalloc instead of dlmalloc in plasma

### PR DESCRIPTION
**Pull request by pcmoritz**
This is currently blocked on https://github.com/jemalloc/jemalloc/issues/1329 but I wanted to post it here too in case anybody has an idea how to move forward.

(@xhochy You are an expert with jemalloc, let me know if you have any ideas on what the right way to do this is. I'd really like to have one memory mapped file only if at all possible so we can remove the sending of file descriptors.)

### CODE REVIEW

File: cpp/src/plasma/CMakeLists.txt [hunk 1]
```diff
@@ -22,6 +22,25 @@ project(plasma VERSION "${ARROW_BASE_VERSION}")
...
+set(JEMALLOC_PREFIX "${CMAKE_CURRENT_BINARY_DIR}/jemalloc_ep-prefix/src/jemalloc_ep/dist/")
+set(jemalloc_URL https://github.com/jemalloc/jemalloc/archive/5.1.0.tar.gz)
+set(JEMALLOC_INCLUDE_DIR "${JEMALLOC_PREFIX}/include")
+set(JEMALLOC_SHARED_LIB "${JEMALLOC_PREFIX}/lib/libjemalloc${CMAKE_SHARED_LIBRARY_SUFFIX}")
+set(JEMALLOC_STATIC_LIB "${JEMALLOC_PREFIX}/lib/libjemalloc_pic${CMAKE_STATIC_LIBRARY_SUFFIX}")
+
+ExternalProject_Add(jemalloc
+    URL ${jemalloc_URL}
+    DOWNLOAD_DIR "${DOWNLOAD_LOCATION}"
+    CONFIGURE_COMMAND ./autogen.sh "--prefix=${JEMALLOC_PREFIX}" "--with-jemalloc-prefix=je_plasma_" "--with-private-namespace=je_plasma_private_" "--disable-tls"
```
Comment (xhochy) [inline cpp/src/plasma/CMakeLists.txt]: Please use the same jemalloc version as we do in Arrow C++. Eventhough they should live happily along eachother, we should not add multiple versions of a library to the project.

Comment (wesm) [reply cpp/src/plasma/CMakeLists.txt]: Aside: if we're able to remove our vendored jemalloc by making this upgrade that would be great

**Comment by wesm** [comment]
Is there a plan about how to proceed on this patch?

**Comment by pcmoritz** [comment]
Yes, the plan is like this:

- Get https://github.com/apache/arrow/pull/3490 merged
- In a follow up PR, impose a memory limit on the number of bytes that are allocated from the plasma side (i.e. plasma will keep track of the memory limit), and if jemalloc will require a bit more due to fragmentation and metadata, that's ok. We will need to make sure that the files are living on a file system that is ~ 10% or so larger than the plasma memory limit to account for the extra memory.
- Don't assume we will be able to coerce jemalloc to use exactly one memory mapped file (that was the original plan, but I've given up on that and with https://github.com/apache/arrow/pull/3124 it's not so important any more),

cc @anuragkh who suggested this plan

**Comment by emkornfield** [comment]
@pcmoritz it looks like this is has gotten fairly out of date, do you have plans to pick it up soon, or should we close the PR?

**Pull request closed without merging**